### PR TITLE
Introduce possibility to choose not to use Joseph Form in KU updater

### PR DESCRIPTION
**Pull request by VinInn**
This PR introduces the possibility to choose at compile time between the Joseph Form (current default) and the short, faster form in KU updater
(
see for instance https://en.wikipedia.org/wiki/Kalman_filter#Simplification_of_the_a_posteriori_error_covariance_formula
)
A comparison of the two forms is also coded in

I have tested and compared both approaches on ttbar 50PU and data 2017 High luminosity including a test with APE increased by a factor 10.

MVA results for pu50 in http://innocent.home.cern.ch/innocent/RelVal/pu50_noJF
(spot the difference!)

in general the Joseph form is indeed more stable:
for instance on ttbar pu50 
on 230M invocation of the updater
~1000 times the fast version produces a matrix with negative elements on the diagonal while the Joseph form NEVER.
The maximal difference between for instance the sqrt of the diagonal values can be very large,
still their rms are pretty small (0.00147328/0.000522762/0.000464533/0.043233/0.0130172)
similar results on data
(0.00214165/0.00168747/0.00222978/0.0768335/0.0321059 rms on data)
Unexpectedly with 10xAPE the difference between Joseph and fast form is smaller...
(rms of difference of sqrt of diagonal on data 0.00118256/0.00038112/0.000281301/0.0215102/0.0083532)

For what concern timing gain in NOT using the Joseph Form 
the KU itself speeds up of about 40% while the total tracking time reduces of  ~2%.

As the default stays the Joseph Form NO regression is expected.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
@VinInn : do I understand correctly that the "KU_JF_TEST" driven part of your additions is only meant to be there for some private local test, and not to be switched on in production? Otherwise, the cout in 
https://github.com/cms-sw/cmssw/pull/22733/files#diff-2a930829c767a8b2a883dd5eb87b8c7bR38
should be moved to something more "thread-safe".

For the rest I can confirm that also in relvals there are more "curv error not pos-def" errors when switching on the faster form

File: TrackingTools/KalmanUpdators/src/KFUpdator.cc [hunk 1]
```diff
@@ -7,6 +7,53 @@
...
+    std::atomic<long long> ifnopd;
+
+    std::atomic<unsigned long long> dtot;
+    std::atomic<unsigned long long> idmm[5];
+    double mm = 0;
+    double dmm[5] = {0};
+    bool ok=true;
+    ~Stat() {
+       double n = 1.e-3/double(dtot);
+       std::cout << "KF " << tot*1.e-9 << "G " << mm <<" "
```
Comment (VinInn) [inline TrackingTools/KalmanUpdators/src/KFUpdator.cc]: this IS thread safe as there is a single instance of Stat (stat)
and the destructor run when the library is "unloaded"

**Comment by VinInn** [comment]
KU_JF_TEST is NOT supposed to be run in production, only by developers.
I do not know a better way to achieve the same result w/o such a local instrumentation

**Comment by perrotta** [comment]
+1
- It allows to use the faster form in KU updater: default remains the Joseph Form
- Also some extra debug code can be enabled at compile time by the developers
- No effect expected on standard sequences and workflows: jenkins tests pass

**Comment by fabiocos** [comment]
+1

**Pull request merged**